device/sycl/src/finding/find_tracks.hpp

```diff
@@ -0,0 +1,502 @@
+        copy.get_size(measurements);
+
+    // Get copy of barcode uniques
+    measurement_collection_types::buffer uniques_buffer{n_measurements,
+                                                        mr.main};
+    copy.setup(uniques_buffer)->wait();
+    measurement_collection_types::device uniques(uniques_buffer);
+
+    measurement_collection_types::device::iterator uniques_end =
+        oneapi::dpl::unique_copy(policy, measurements.ptr(),
+                                 measurements.ptr() + n_measurements,
+                                 uniques.begin(), measurement_equal_comp());
+    const unsigned int n_modules =
+        static_cast<unsigned int>(uniques_end - uniques.begin());
+
+    // Get upper bounds of unique elements
+    vecmem::data::vector_buffer<unsigned int> upper_bounds_buffer{n_modules,
+                                                                  mr.main};
+    copy.setup(upper_bounds_buffer)->wait();
+    vecmem::device_vector<unsigned int> upper_bounds(upper_bounds_buffer);
+
+    oneapi::dpl::upper_bound(policy, measurements.ptr(),
+                             measurements.ptr() + n_measurements,
+                             uniques.begin(), uniques.begin() + n_modules,
+                             upper_bounds.begin(), measurement_sort_comp());
+    queue.wait_and_throw();
+
+    /*****************************************************************
+     * Kernel1: Create barcode sequence
+     *****************************************************************/
+
+    vecmem::data::vector_buffer<detray::geometry::barcode> barcodes_buffer{
+        n_modules, mr.main};
+    copy.setup(barcodes_buffer)->wait();
+
+    queue
+        .submit([&](::sycl::handler& h) {
+            h.parallel_for<
+                typename kernels_t::make_barcode_sequence_kernel_type>(
+                calculate1DimNdRange(n_modules, 64),
+                [uniques_view = vecmem::get_data(uniques_buffer),
+                 barcodes_view = vecmem::get_data(barcodes_buffer)](
+                    ::sycl::nd_item<1> item) {
+                    device::make_barcode_sequence(
+                        details::global_index(item),
+                        {uniques_view, barcodes_view});
+                });
+        })
+        .wait_and_throw();
+
+    const unsigned int n_seeds = copy.get_size(seeds);
+
+    // Prepare input parameters with seeds
+    bound_track_parameters_collection_types::buffer in_params_buffer(n_seeds,
+                                                                     mr.main);
+    copy.setup(in_params_buffer)->wait();
+    copy(seeds, in_params_buffer, vecmem::copy::type::device_to_device)->wait();
+    vecmem::data::vector_buffer<unsigned int> param_liveness_buffer(n_seeds,
+                                                                    mr.main);
+    copy.setup(param_liveness_buffer)->wait();
+    copy.memset(param_liveness_buffer, 1)->wait();
```

Comment:
Make of these operations can be performed concurrently is we launch them asynchronously; I'd suggest we start using the event API that SYCL provides (as in e.g. device/sycl/src/sanity/contiguous_on.hpp) as the easy of its task graph building APIs is one of the biggest strengths of SYCL.

Reply:
As I wrote, I'm very open to exploring all of these. Not in this PR though. I decidedly made the choice of just waiting for everything in this first version.